include/libcyphal/platform/single_threaded_executor.hpp

```diff
@@ -113,114 +98,176 @@ class SingleThreadedExecutor : public IExecutor  // NOSONAR cpp:S4963
-        ScheduledNode& operator=(const ScheduledNode&)     = delete;
-        ScheduledNode& operator=(ScheduledNode&&) noexcept = delete;
+        CallbackNode(SingleThreadedExecutor* const executor, Callback::Function&& function)
+            : executor_{executor}
+            , function_{std::move(function)}
+            , next_exec_time_{TimePoint::max()}
+            , schedule_{Callback::Schedule::None{}}
+        {
+            CETL_DEBUG_ASSERT(executor_, "");
+            CETL_DEBUG_ASSERT(function_, "");
+            executor_->onCallbackHandling({}, static_cast<Callback::Handle>(*this));
+        }
+        ~CallbackNode()
+        {
+            if (auto** const root_node_ptr = getRootNodePtr())
+            {
+                remove(*root_node_ptr, this);
+            }
+            executor_->onCallbackHandling(static_cast<Callback::Handle>(*this), {});
+        };
+
+        CallbackNode(CallbackNode&& other) noexcept
+            : Node(std::move(static_cast<Node&>(other)))
+            , executor_{other.executor_}
+            , function_{std::move(other.function_)}
+            , next_exec_time_{other.next_exec_time_}
+            , schedule_{other.schedule_}
+        {
+            executor_->onCallbackHandling(static_cast<Callback::Handle>(other), static_cast<Callback::Handle>(*this));
+        }
+
+        CallbackNode& operator=(CallbackNode&& other) noexcept
+        {
+            static_cast<Node&>(*this) = std::move(static_cast<Node&>(other));
+
+            executor_       = other.executor_;
+            function_       = std::move(other.function_);
+            next_exec_time_ = other.next_exec_time_;
+            schedule_       = other.schedule_;
 
-        cetl::optional<Callback::Schedule::Variant>& schedule() noexcept
+            executor_->onCallbackHandling(static_cast<Callback::Handle>(other), static_cast<Callback::Handle>(*this));
+
+            return *this;
+        }
+
+        CallbackNode(const CallbackNode&)            = delete;
+        CallbackNode& operator=(const CallbackNode&) = delete;
+
+        explicit operator Callback::Handle() const noexcept
+        {
+            // Next nolint & NOSONAR are unavoidable: this is opeque handle coversion code.
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            return reinterpret_cast<Callback::Handle>(this);  // NOSONAR cpp:S3630
+        }
+        static CallbackNode* fromHandle(const Callback::Handle handle) noexcept
+        {
+            CETL_DEBUG_ASSERT(handle, "");
+
+            // Next nolint & NOSONAR are unavoidable: this is opeque handle coversion code.
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast, performance-no-int-to-ptr)
+            return reinterpret_cast<CallbackNode*>(handle);  // NOSONAR cpp:S3630
+        }
+
+        SingleThreadedExecutor* executor() const noexcept
+        {
+            return executor_;
```

Comment:
can this be null? If not can it be a reference?

Reply:
It can't be null, and it can't be (unfortunately) a reference - otherwise it will be impossible to have move assignment operator. But I believe I can make it as a reference in constructor @ line 160 (currently it is a pointer with assert there) - just to emphasize that it can't null.

Reply:
Actually, it's possible - see last commit with `std::reference_wrapper`.